titan/features/incar.py

```diff
@@ -0,0 +1,171 @@
+                            self.agent.mean_num_partners[bond], model.np_random
+                        )
+                    model.pop.update_partnerability(self.agent)
+
+                # does agent stay on haart
+                if hiv_bool:
+                    if self.agent.haart.active:
+                        if (
+                            model.run_random.random()
+                            <= self.agent.location.params.incar.haart.discontinue
+                        ):  # 12% remain surpressed
+                            self.agent.haart.active = False
+                            self.agent.haart.adherence = 0
+
+                        # END FORCE
+
+        # should the agent become incarcerated?
+        elif model.run_random.random() < (
+            self.agent.location.params.demographics[self.agent.race][
+                self.agent.sex_type
+            ].incar.prob
+            * hiv_multiplier
+            * model.calibration.incarceration
+        ):
+            incar_duration = self.agent.location.params.demographics[self.agent.race][
+                self.agent.sex_type
+            ].incar.duration.prob
+
+            bin = current_p_value = 1
+            p = model.run_random.random()
+            while p >= current_p_value:
+                current_p_value += incar_duration[bin].prob
+                bin += 1
+
+            timestay = model.run_random.randint(
+                incar_duration[bin].min, incar_duration[bin].max
+            )
+
+            self.active = True
+            self.time = timestay
+
+            if hiv_bool:
+                if not self.agent.hiv_dx:
+                    if (
+                        model.run_random.random()
+                        < self.agent.location.params.incar.hiv.dx
+                    ):
+                        self.agent.hiv_dx = True
+                else:  # Then tested and HIV, check to enroll in ART
+                    if (
+                        model.run_random.random()
+                        < self.agent.location.params.incar.haart.prob
+                    ):
+                        if (
+                            model.run_random.random()
+                            < self.agent.location.params.incar.haart.adherence
+                        ):
+                            adherence = 5
+                        else:
+                            adherence = model.run_random.randint(1, 4)
+
+                        # Add agent to HAART class set, update agent params
+                        self.agent.haart.active = True
+                        self.agent.haart.adherence = adherence
+                        self.agent.haart.time = model.time
+
+            # PUT PARTNERS IN HIGH RISK
+            for bond in self.agent.location.params.high_risk.partnership_types:
+                for partner in self.agent.partners[bond]:
+                    if not partner.high_risk.active and model.params.features.high_risk:
```

Comment:
Probably faster to check if high_risk is turned on in features before initiating the loop